x/incentive/msgs.go

```diff
@@ -0,0 +1,257 @@
+		return err
+	}
+	if err := validateProposedIncentiveProgram(msg.Program); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetSignBytes implements Msg
+func (msg MsgGovCreateProgram) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// GetSigners implements Msg
+func (msg MsgGovCreateProgram) GetSigners() []sdk.AccAddress {
+	return checkers.Signers(msg.Authority)
+}
+
+func NewMsgCreateAndSponsorProgram(authority, title, description, sponsor string, program IncentiveProgram,
+) *MsgGovCreateAndSponsorProgram {
+	return &MsgGovCreateAndSponsorProgram{
+		Title:       title,
+		Description: description,
+		Program:     program,
+		Authority:   authority,
+		Sponsor:     sponsor,
+	}
+}
+
+func (msg MsgGovCreateAndSponsorProgram) ValidateBasic() error {
+	if err := checkers.ValidateProposal(msg.Title, msg.Description, msg.Authority); err != nil {
+		return err
+	}
+	if _, err := sdk.AccAddressFromBech32(msg.Sponsor); err != nil {
+		return sdkerrors.Wrap(err, "invalid sponsor address")
+	}
+	if err := validateProposedIncentiveProgram(msg.Program); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetSignBytes implements Msg
+func (msg MsgGovCreateAndSponsorProgram) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// GetSigners implements Msg
+func (msg MsgGovCreateAndSponsorProgram) GetSigners() []sdk.AccAddress {
+	return checkers.Signers(msg.Authority)
+}
+
+// validateProposedIncentiveProgram runs IncentiveProgram.Validate and also checks additional requirements applying
+// to incentive programs which have not yet been funded or passed by governance
+func validateProposedIncentiveProgram(program IncentiveProgram) error {
+	if program.Id != 0 {
+		return ErrInvalidProgramID.Wrapf("%d", program.Id)
+	}
+	if !program.RemainingRewards.IsZero() {
+		return ErrNonzeroRemainingRewards.Wrap(program.RemainingRewards.String())
+	}
+	if !program.FundedRewards.IsZero() {
+		return ErrNonzeroFundedRewards.Wrap(program.FundedRewards.String())
+	}
+	return program.Validate()
+}
+
+func validateSenderAsset(sender string, asset *sdk.Coin) error {
+	_, err := sdk.AccAddressFromBech32(sender)
+	if err != nil {
+		return err
+	}
+	if asset == nil {
+		return ErrNilAsset
+	}
+	if err := asset.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validateSenderAssetTier(sender string, tier uint32, asset *sdk.Coin) error {
+	if err := validateSenderAsset(sender, asset); err != nil {
+		return err
+	}
+	if tier < 1 || tier > 3 {
+		return ErrInvalidTier.Wrapf("%d", tier)
+	}
+	return nil
+}
+
+// Validate performs validation on an IncentiveProgram type returning an error
+// if the program is invalid.
+func (ip IncentiveProgram) Validate() error {
+	if err := sdk.ValidateDenom(ip.Denom); err != nil {
+		return err
+	}
+	if !strings.HasPrefix(ip.Denom, "u/") {
```

Comment:
Don't we have a function for that in `x/leverage`? If not, let's create a common function, eg `assertIsUToken`

Reply:
We have one - originally, I was worried about an import cycle between leverage an incentive types and did not want to use `leveragetypes.HasUTokenPrefix`. I'll use it for now, since a cycle doesn't seem likely yet.